include/deal.II/base/hdf5.h

```diff
@@ -0,0 +1,712 @@
+                    const std::vector<hsize_t> &offset,
+                    const std::vector<hsize_t> &stride,
+                    const std::vector<hsize_t> &count,
+                    const std::vector<hsize_t> &block);
+
+    /**
+     * This function does not write any data, but it can contribute to a
+     * collective write call. Number can be `float`, `double`,
+     * `std::complex<float>`, `std::complex<double>`, `int` or `unsigned int`.
+     *
+     * Datatype conversion takes place at the time of a read or write and is
+     * automatic. See the <a
+     * href="https://support.hdfgroup.org/HDF5/doc/UG/HDF5_Users_Guide-Responsive%20HTML5/index.html#t=HDF5_Users_Guide%2FDatatypes%2FHDF5_Datatypes.htm%23TOC_6_10_Data_Transferbc-26&rhtocid=6.5_2">Data
+     * Transfer: Datatype Conversion and Selection</a>  section in the HDF5
+     * User's Guide.
+     */
+    template <typename number>
+    void
+    write_none();
+
+    /**
+     * This funcion retrieves the type of I/O that was performed on the last
+     * parallel I/O call. See <a
+     * href="https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioActualIoMode">H5Pget_mpio_actual_io_mode</a>.
+     * The return type T can be H5D_mpio_actual_io_mode_t or std::string. The
+     * type H5D_mpio_actual_io_mode_t corresponds to the value returned by
+     * H5Pget_mpio_actual_io_mode and std::string is a human readable
+     * conversion.
+     */
+    template <typename T>
+    T
+    io_mode();
+
+    /**
+     * This funcion retrieves the local causes that broke collective I/O on the
+     * last parallel I/O call. See <a
+     * href="https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioNoCollectiveCause">H5Pget_mpio_no_collective_cause</a>.
+     * The return type T can be uint32_t or std::string. The type uint32_t
+     * corresponds to the value returned by H5Pget_mpio_no_collective_cause and
+     * std::string is a human readable conversion.
+     */
+    template <typename T>
+    T
+    local_no_collective_cause();
+
+    /**
+     * This funcion retrieves the global causes that broke collective I/O on the
+     * last parallel I/O call. See <a
+     * href="https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioNoCollectiveCause">H5Pget_mpio_no_collective_cause</a>.
+     * The return type T can be uint32_t or std::string. The type uint32_t
+     * corresponds to the value returned by H5Pget_mpio_no_collective_cause and
+     * std::string is a human readable conversion.
+     */
+    template <typename T>
+    T
+    global_no_collective_cause();
+
+    /**
+     * This function retrieves the IO mode checking. If check_io_mode is true,
+     * then after every read and write operation in the dataset, it will be
+     * retrieved the type of I/O that was performed on the last parallel I/O
+     * call If check_io_mode is false then no checking will be performed.
```

Comment:
I added the comment:
The default access mode in the HDF5 C++ interface of deal.ii is collective which is typically faster since it allows MPI to do more optimizations. In some cases, such as when there is type conversion, the HDF5 library can decide to do independent I/O instead of collective I/O, even if the user asks for collective I/O. See the following [article](https://www.hdfgroup.org/2015/08/parallel-io-with-hdf5/)

In cases where maximum performance is a requirement, it is important to make sure that all MPI read/write operations are collective. The HDF5 library provides API routines that can be used after the read/write I/O operations to query the I/O mode. If DataSet::query_io_mode() is set to True, then after every read/write operation the HDF5 deal.ii interface calls the routines [H5Pget_mpio_actual_io_mode()](https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioActualIoMode) and [H5Pget_mpio_no_collective_cause()](https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioNoCollectiveCause). The results are stored in `DataSet::io_mode`, `DataSet::local_no_collective_cause` and `DataSet::get_global_no_collective_cause`. We suggest to query the I/O mode only in Debug mode because it requires calling additional HDF5 routines.